src/main/java/uk/nhs/digital/nhsconnect/nhais/service/RecepConsumerService.java

```diff
@@ -3,67 +3,119 @@
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.dao.DuplicateKeyException;
 import org.springframework.stereotype.Component;
-import uk.nhs.digital.nhsconnect.nhais.model.edifact.Recep;
+import uk.nhs.digital.nhsconnect.nhais.model.edifact.Interchange;
+import uk.nhs.digital.nhsconnect.nhais.model.edifact.Message;
 import uk.nhs.digital.nhsconnect.nhais.model.edifact.ReferenceMessageRecep;
 import uk.nhs.digital.nhsconnect.nhais.model.mesh.MeshMessage;
-import uk.nhs.digital.nhsconnect.nhais.parse.RecepParser;
+import uk.nhs.digital.nhsconnect.nhais.model.mesh.WorkflowId;
+import uk.nhs.digital.nhsconnect.nhais.parse.EdifactParser;
 import uk.nhs.digital.nhsconnect.nhais.repository.InboundState;
 import uk.nhs.digital.nhsconnect.nhais.repository.InboundStateRepository;
 import uk.nhs.digital.nhsconnect.nhais.repository.OutboundStateRepository;
 import uk.nhs.digital.nhsconnect.nhais.repository.OutboundStateRepositoryExtensions;
 
 import java.time.Instant;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
 
 @Component
 @Slf4j
 @RequiredArgsConstructor(onConstructor = @__(@Autowired))
 public class RecepConsumerService {
 
-    private final RecepParser recepParser;
+    private final EdifactParser edifactParser;
     private final OutboundStateRepository outboundStateRepository;
     private final InboundStateRepository inboundStateRepository;
 
     public void handleRecep(MeshMessage meshMessage) {
         LOGGER.info("Received RECEP message: {}", meshMessage);
-        Recep recep = recepParser.parse(meshMessage.getContent());
+        var recep = edifactParser.parse(meshMessage.getContent());
         LOGGER.debug("Parsed RECEP message into: {}", recep);
 
-        var inboundState = InboundState.fromRecep(recep);
-        if (!saveState(inboundState)) {
-            return;
-        }
+        var messagesToProcess = filterOutDuplicates(recep);
 
+        var outboundStateUpdates = prepareOutboundStateParams(messagesToProcess);
+        var inboundStateInserts = prepareInboundStateInserts(messagesToProcess);
```

Comment:
(Optional) You could refactor these to "list classes" to be able to test this code:
````java
    private class OutboundStateParams extends AbstractList<OutboundStateRepositoryExtensions.UpdateRecepParams> {
        private final List<OutboundStateRepositoryExtensions.UpdateRecepParams> params;

        public OutboundStateParams(List<Message> messages) {
            this.params = prepareOutboundStateParams(messages);
        }

        private List<OutboundStateRepositoryExtensions.UpdateRecepParams> prepareOutboundStateParams(List<Message> messagesToProcess) {
            return messagesToProcess.stream()
                    .map(this::prepareOutboundStateUpdateParams)
                    .flatMap(Collection::stream)
                    .collect(Collectors.toList());
        }

        private List<OutboundStateRepositoryExtensions.UpdateRecepParams> prepareOutboundStateUpdateParams(Message message) {
            LOGGER.info("Handling recep message {}", message);
            //sender is swapped with recipient as communication is done the opposite way
            var outbound_sender = message.getInterchange().getInterchangeHeader().getRecipient();
            var outbound_recipient = message.getInterchange().getInterchangeHeader().getSender();
            var dateTimePeriod = message.getTranslationDateTime().getTimestamp();
            var interchangeSequence = message.getReferenceInterchangeRecep().getInterchangeSequenceNumber();

            return message.getReferenceMessageReceps().stream()
                    .map(referenceMessageRecep -> prepareOutboundStateUpdateParams(outbound_sender, outbound_recipient, dateTimePeriod, interchangeSequence, referenceMessageRecep))
                    .collect(Collectors.toList());
        }

        private OutboundStateRepositoryExtensions.UpdateRecepParams prepareOutboundStateUpdateParams(
                String outbound_sender, String outbound_recipient, Instant dateTimePeriod, Long interchangeSequence, ReferenceMessageRecep referenceMessageRecep) {

            var messageSequence = referenceMessageRecep.getMessageSequenceNumber();
            var recepCode = referenceMessageRecep.getRecepCode();

            return new OutboundStateRepositoryExtensions.UpdateRecepParams(
                    outbound_sender, outbound_recipient, interchangeSequence, messageSequence, recepCode, dateTimePeriod);
        }
````

Reply:
Good, but I leave it as is now